File: July2021/DecodeWaysII.java
public class Solution {
  // TC : O(n)
  // SC : O(n)
  public int numDecodings(String s) {

    long[] dp = new long[s.length()+1];
    dp[0] = 1;
    dp[1] = decodeChar(s.charAt(0));

    for (int i=2; i<=s.length(); i++) {

      char prevChar = s.charAt(i-2);
      char currChar = s.charAt(i-1);

      // Single char decoding
      dp[i] += dp[i-1] * decodeChar(currChar);

      // Double char decoding
      dp[i] += dp[i-2] * decodeTwoChars(prevChar, currChar);

      dp[i] = dp[i]%1000000007;
    }
    return (int)dp[s.length()];
  }

  private int decodeChar(char c) {
    if (c == '*') {
      return 9;
    } else if (c == '0') {
      return 0;
    }
    return 1;
  }

  private int decodeTwoChars(char first, char second) {
    switch (first) {
      case '*':
        if (second == '*') {
          return 15;
        } else if (second >= '0' && second <= '6') {
          return 2;
        } else {
          return 1;
        }
      case '1':
        if (second == '*') {
          return 9;
        } else {
          return 1;
        }
      case '2':
        if (second == '*') {
          return 6;
        } else if (second >= '0' && second <= '6') {
          return 1;
        } else {
          return 0;
        }
      default:
        return 0;
    }
  }
}

File: July2021/C++/DecodeWaysII.cpp
class Solution
{
public:
    bool betweenJZ(char l, char r)
    {
        if (l == '1' && r >= '0' && r <= '9')
            return true;
        if (l == '2' && r >= '0' && r <= '6')
            return true;
        return false;
    }

    int numDecodings(string s)
    {

        if (s[0] == '0')
            return 0;
        int MOD = 1e9 + 7;
        vector<long long> dp(s.length() + 1, 0);
        dp[1] = dp[0] = 1;
        if (s[0] == '*')
            dp[1] = 9;
        for (int i = 2; i < dp.size(); ++i)
        {
            if (s[i - 1] == '0')
            {
                if (s[i - 2] >= '3' || s[i - 2] == '0')
                    return 0;
                dp[i] = dp[i - 2] % MOD;
                if (s[i - 2] == '*')
                    dp[i] = (dp[i] * 2) % MOD;
            }
            else if (betweenJZ(s[i - 2], s[i - 1]))
            {
                dp[i] = (dp[i - 1] + dp[i - 2]) % MOD;
            }
            else if (s[i - 1] == '*')
            {
                dp[i] = (dp[i - 1] * 9) % MOD;
                if (s[i - 2] == '1' || s[i - 2] == '*')
                    dp[i] += (dp[i - 2] * 9) % MOD;
                if (s[i - 2] == '2' || s[i - 2] == '*')
                    dp[i] += (dp[i - 2] * 6) % MOD;
            }
            else
            {
                dp[i] = dp[i - 1] % MOD;
                if (s[i - 2] == '*')
                {
                    dp[i] = (dp[i] + dp[i - 2]) % MOD;
                    if (s[i - 1] <= '6')
                        dp[i] = (dp[i] + dp[i - 2]) % MOD;
                }
            }
        }
        return dp[dp.size() - 1] % MOD;
    }
};